File: typescript/tests/core/rotate-key.test.ts
/**
 * License-key rotation — leak response.
 *
 * Rotation replaces the license_key AND revokes every active seat. Both
 * halves matter: a rotation that left existing devices running would not
 * contain a leaked key, which is its only purpose.
 *
 * Mirrors Go's TestRotateLicenseKey_* in licensing/rotate_key_test.go.
 */

import { Database } from 'bun:sqlite';
import { describe, expect, it } from 'bun:test';

import { MemoryStorage } from '@anorebel/licensing/storage/memory';
import { SqliteStorage } from '@anorebel/licensing/storage/sqlite';
import { applyMigrations as applySqliteMigrations } from '@anorebel/licensing/storage/sqlite/migrations';

import { createFixedClock } from '../../src/id.ts';
import { generateLicenseKey } from '../../src/license-key.ts';
import { createLicense, rotateLicenseKey } from '../../src/license-service.ts';
import { revoke } from '../../src/lifecycle.ts';
import type { Storage } from '../../src/storage/types.ts';
import { registerUsage } from '../../src/usage-service.ts';

const NOW = '2026-06-01T00:00:00.000000Z';
const clock = createFixedClock(NOW);

interface Backend {
  readonly name: string;
  readonly make: () => Promise<{ s: Storage; cleanup: () => void }>;
}

const BACKENDS: readonly Backend[] = [
  {
    name: 'memory',
    make: async () => ({ s: new MemoryStorage({ clock }), cleanup: () => undefined }),
  },
  {
    name: 'sqlite',
    make: async () => {
      const db = new Database(':memory:');
      applySqliteMigrations(db);
      return {
        s: new SqliteStorage(db, { clock, skipWalPragma: true }),
        cleanup: () => db.close(),
      };
    },
  },
];

async function newLicense(s: Storage) {
  return createLicense(
    s,
    clock,
    {
      scope_id: null,
      template_id: null,
      licensable_type: 'User',
      licensable_id: 'rot-1',
      license_key: generateLicenseKey(),
      max_usages: 5,
    },
    {},
  );
}

for (const backend of BACKENDS) {
  describe(`rotateLicenseKey (${backend.name})`, () => {
    it('replaces the key and revokes every seat', async () => {
      const { s, cleanup } = await backend.make();
      try {
        const license = await newLicense(s);
        const oldKey = license.license_key;

        for (const prefix of ['a', 'b']) {
          await registerUsage(s, clock, {
            license_id: license.id,
            fingerprint: `${prefix}1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2`,
          });
        }

        const res = await rotateLicenseKey(s, clock, license.id, {
          actor: 'admin:rotate-key',
          actorKind: 'admin',
        });

        expect(res.license.license_key).not.toBe(oldKey);
        expect(res.revokedUsageIds).toHaveLength(2);

        // The old key must stop resolving; the new one must start.
        expect(await s.getLicenseByKey(oldKey)).toBeNull();
        expect(await s.getLicenseByKey(res.license.license_key)).not.toBeNull();

        // No live seats remain, so every device must re-activate.
        const page = await s.listUsages(
          { license_id: license.id, status: ['active'] },
          { limit: 10 },
        );
        expect(page.items).toHaveLength(0);
      } finally {
        cleanup();
      }
    });

    it('records the rotation without writing either key to the audit log', async () => {
      const { s, cleanup } = await backend.make();
      try {
        const license = await newLicense(s);
        const oldKey = license.license_key;

        const res = await rotateLicenseKey(s, clock, license.id, {
          actor: 'admin:rotate-key',
          actorKind: 'admin',
        });

        const audit = await s.listAudit({ event: 'license.key_rotated' }, { limit: 10 });
        expect(audit.items).toHaveLength(1);
        expect(audit.items[0]?.actor_kind).toBe('admin');

        // An append-only log that support staff can read must not carry the
        // secret the rotation exists to protect.
        const serialised = JSON.stringify([audit.items[0]?.prior_state, audit.items[0]?.new_state]);
        expect(serialised).not.toContain(oldKey);
        expect(serialised).not.toContain(res.license.license_key);
      } finally {
        cleanup();
      }
    });

    it('refuses to rotate a revoked license', async () => {
      const { s, cleanup } = await backend.make();
      try {
        const license = await newLicense(s);
        await s.withTransaction(async (tx) => {
          const cur = await tx.getLicense(license.id);
          if (cur !== null) await revoke(tx, cur, clock, {});
        });

        // Revoked is terminal; a new key would imply it could still be used.
        await expect(rotateLicenseKey(s, clock, license.id, {})).rejects.toMatchObject({
          code: 'LicenseRevoked',
        });
      } finally {
        cleanup();
      }
    });
  });
}

File: licensing/rotate_key_test.go
package licensing_test

import (
	"testing"

	lic "github.com/AnoRebel/licensing/licensing"
	"github.com/AnoRebel/licensing/licensing/storage/memory"
)

type rotClock struct{ now string }

func (c rotClock) NowISO() string { return c.now }

func newRotFixture(t *testing.T) (*memory.Storage, rotClock, *lic.License) {
	t.Helper()
	clk := rotClock{now: "2026-06-01T00:00:00.000000Z"}
	st := memory.New(memory.Options{Clock: clk})
	l, err := lic.CreateLicense(st, clk, lic.CreateLicenseInput{
		LicensableType: "User", LicensableID: "rot-1",
		LicenseKey: lic.GenerateLicenseKey(),
		Status:     lic.LicenseStatusActive, MaxUsages: 5,
	}, lic.CreateLicenseOptions{})
	if err != nil {
		t.Fatal(err)
	}
	return st, clk, l
}

// Rotation must change the key AND clear every seat — a rotation that left
// devices running would not contain a leaked key, which is its only purpose.
func TestRotateLicenseKey_ReplacesKeyAndRevokesSeats(t *testing.T) {
	st, clk, l := newRotFixture(t)
	oldKey := l.LicenseKey

	for _, fp := range []string{"a", "b"} {
		if _, err := lic.RegisterUsage(st, clk, lic.RegisterUsageInput{
			LicenseID: l.ID, Fingerprint: fp + "1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2",
		}, lic.RegisterUsageOptions{}); err != nil {
			t.Fatal(err)
		}
	}

	res, err := lic.RotateLicenseKey(st, clk, l.ID, lic.RotateLicenseKeyOptions{
		Actor: "admin:rotate-key", ActorKind: lic.ActorAdmin,
	})
	if err != nil {
		t.Fatal(err)
	}

	if res.License.LicenseKey == oldKey {
		t.Fatal("license_key did not change")
	}
	if len(res.RevokedUsageIDs) != 2 {
		t.Fatalf("expected 2 seats revoked, got %d", len(res.RevokedUsageIDs))
	}

	// The old key must no longer resolve; the new one must.
	if got, _ := st.GetLicenseByKey(oldKey); got != nil {
		t.Fatal("the old license_key still resolves")
	}
	if got, _ := st.GetLicenseByKey(res.License.LicenseKey); got == nil {
		t.Fatal("the new license_key does not resolve")
	}

	// Every seat is gone, so each device must re-activate.
	page, err := st.ListUsages(lic.LicenseUsageFilter{
		LicenseID: &l.ID, Status: []lic.UsageStatus{lic.UsageStatusActive},
	}, lic.PageRequest{Limit: 10})
	if err != nil {
		t.Fatal(err)
	}
	if len(page.Items) != 0 {
		t.Fatalf("expected no active seats after rotation, got %d", len(page.Items))
	}
}

// The audit row records that a rotation happened without recording either
// key — an append-only log that support staff can read must not carry the
// secret the rotation exists to protect.
func TestRotateLicenseKey_AuditOmitsKeyMaterial(t *testing.T) {
	st, clk, l := newRotFixture(t)
	oldKey := l.LicenseKey

	res, err := lic.RotateLicenseKey(st, clk, l.ID, lic.RotateLicenseKeyOptions{
		Actor: "admin:rotate-key", ActorKind: lic.ActorAdmin,
	})
	if err != nil {
		t.Fatal(err)
	}

	page, err := st.ListAudit(lic.AuditLogFilter{
		Events: []string{"license.key_rotated"},
	}, lic.PageRequest{Limit: 10})
	if err != nil {
		t.Fatal(err)
	}
	if len(page.Items) != 1 {
		t.Fatalf("expected 1 license.key_rotated row, got %d", len(page.Items))
	}
	row := page.Items[0]
	if row.ActorKind != lic.ActorAdmin {
		t.Fatalf("actor_kind = %q, want admin", row.ActorKind)
	}
	for _, state := range []map[string]any{row.PriorState, row.NewState} {
		for k, v := range state {
			if s, ok := v.(string); ok && (s == oldKey || s == res.License.LicenseKey) {
				t.Fatalf("audit field %q leaked a license key", k)
			}
		}
	}
}

// Revoked is terminal; handing out a new key would imply the license could
// still be used.
func TestRotateLicenseKey_RefusesRevokedLicense(t *testing.T) {
	st, clk, l := newRotFixture(t)
	if err := st.WithTransaction(func(tx lic.StorageTx) error {
		cur, err := tx.GetLicense(l.ID)
		if err != nil {
			return err
		}
		_, err = lic.Revoke(tx, cur, clk, lic.TransitionOptions{})
		return err
	}); err != nil {
		t.Fatal(err)
	}

	if _, err := lic.RotateLicenseKey(st, clk, l.ID, lic.RotateLicenseKeyOptions{}); err == nil {
		t.Fatal("expected rotation of a revoked license to fail")
	}
}
